### DISCUSSION TITLE
Leveraging the free 5GB iCloud storage

### DISCUSSION THREAD

**Discussion by CreativePR** [icloud-photos-downloader/icloud_photos_downloader#795] [General]
Many of the iOS clients such as Nextcloud work horribly for backgroup uploads.  If you read on the iOS github for the Nextcloud IOS client, you'll see pages and pages of problems...  Basically, the client will not upload in the background properly due to bogus Apple limitations.  I've heard that some apps like Amazon photos and Google photos work better but there are some users who do not want to pay for those services either.

Been thinking to leverage the free 5GB storage that every iPhone owner has to use for the background uploading.  In this scenerio say we have a user who has their 5gB free storage and does not want to pay Apple for a paid tier level.  On their phone they have say 200GB of photos and videos they want to sync with their nextcloud account painlessly without connecting to a computer, using the horrible nextcloud IOS app, etc.  They want them to magically sync in the background.

User
Phone using as normal, taking pics, etc.
Phone is uploading to their free 5GB iCLoud storage
ICPD pulls newly uploaded photos and sends them to the user's nextcloud account.  As each photo is downloaded, command is sent to delete it from user's iCloud account - this way the 5gb storage never fills up.

Problem, if command is sent to delete from iCloud by ICPD, then the photo is also deleted from the user's iCloud photo gallery.  Many users do not want this behavior.  They want the photos to remain in their photo gallery until they choose to delete them even if they have been deleted from their iCloud account.  The only want to do this is to disable iCloud uploading which defeats the whole process.

Any ideas of a way around this?  If the nextcloud IOS client actually worked correctly for background uploading of photos and videos, this would not be an issue but it seems that the only practical way to have a guaranteed background upload of photos and videos by the user's iPhone is to go through iCloud and then use ICPD to sync with their Nextcloud account.

**Comment by AndreyNikiforov** [comment]
> Problem, if command is sent to delete from iCloud by ICPD, then the photo is also deleted from the user's iCloud photo gallery. Many users do not want this behavior. They want the photos to remain in their photo gallery until they choose to delete them even if they have been deleted from their iCloud account.

I am not following the logic here. If users want to keep Apple iCloud free tier, they need to keep cleaning it. Either automatically with icloudpd `--delete-after-download` parameter or manually while running icloudpd without `--auto-delete` parameter.

I am not familiar with nextcloud. If it has capability to mount storage as local folder, than that folder can be specified as destination for icloudpd.

**Reply by CreativePR** [reply]
Ok, let me restate.

Facts
- Nextcloud iOS app doesn't work properly or consistently for background uploading of iPhone user's photos/videos from their camera roll. Looking at the GitHub page for the app, it appears this is a long ongoing problem that the developer either cannot or will not fix.  It appears related to issues with iOS limits with background operations of third party iOS apps. Company's such as Google and Amazon appear to have possibly worked around this issue with their apps but they have only offer small "free" storage tiers or paid tier (Amazon required a prime membership)for "both" photos and videos with "Live Photo" support. The Nextcloud iOS app developer has not. User's of the app have a less than optimum user experience and must constantly fiddle with the app - sometimes with long wait times to upload their camera roll to their Nextcloud account. Overall, it is a very bad user experience. 

- iPhone users receive 5GB of free iCloud storage. Users can opt to sync their camera roll with iCloud. iOS will seamlessly upload in the background which provides an optimum user experience. If an iPhone user has a large (more than 5GB) camera roll, and has opted in to sync their camera roll with iCloud, their free 5GB tier of iCloud storage will quickly fill up and the iOS will throw constant messages that the user needs to upgrade to a paid iCloud tier and some iPhone functions might start to not work or sync properly. These are nag messages and provide a bad user experience.  The only way to stop the messages is the user must upgrade to a paid iCloud tier or disable iCloud sync functionality. 

- My understanding is when a photo/video is deleted from iCloud, it is also deleted from the user's camera roll if they are currently syncing their iPhone camera roll with iCloud

- Photos/Videos downloaded from iCloud using ICPD can be synced with a user's Nextcloud account using customized code


Scenario 
- A group of iPhone users who have Nextcloud accounts

- These user's have only the free 5GB storage tier of iCloud from Apple. They do not want to pay for a paid storage tier

- These users have more than 5GB of photos and videos in their iPhone camera roll. If they enable syncing of their camera roll with iCloud, the camera roll will seamlessly upload to their iCloud account and quickly fill their free 5GB of storage. 

- These users do not want photos/videos deleted from their iPhone camera roll until they choose to delete them manually in the iOS camera roll

- These users want their camera roll synced with their Nextcloud account seamlessly in the background the same as it would be using the iCloud camera roll sync function. 

-  These users "do not" want to use another (paid) service such as Google, Amazon or other paid iOS apps to sync their camera rolls in the background which could then be synced to their Nextcloud account. 


Idea
- Since iOS can sync the camera roll in the background with iCloud, trying to think of a way to leverage the free iCloud storage tier to accomplish syncing in the background with the user's photos/videos with their Nextcloud account.

- We will use ICPD (and custom code) to sync the user's iCloud photo/videos with their Nextcloud account. When a photo/video is downloaded with ICPD, command will be sent to delete that photo/video from iCloud (to keep the storage from exceeding the 5GB free tier limit). 

Problem to solve
- if command is sent to delete from iCloud by ICPD, then the photo is also deleted from the user's iCloud camera roll. As stated above our users do not want this behavior. They want the photos to remain in their camera roll until they choose to delete them even if they have been deleted from their iCloud account.

**Reply by boredazfcuk** [reply]
Has this been raised in the wrong repo? My container has Nextcloud synchronisation capability.

Edit: it also offers the behaviour described in "problems to solve"

**Reply by CreativePR** [reply]
Hi, I've seen your container but haven't had a chance to test it.  I've wrote some custom code that uses ICPD to synchronize (on a basic level) with the user's iCloud account.  I'm sure from what I read of yours, it probably does this alot better.  My code at this point was just for testing/exploratory purposes.

Can you please explain in a bit more detail how your container offers "the behavior described in" my problems to solve section?

Thank You

**Reply by boredazfcuk** [reply]
It syncs to Nextcloud, but you have the option to turn off deletes.

With deletes disabled the workflow is:

1) take photo
2) photo syncs to iCloud.com
3) photo downloaded by icloupd to storage mounted in my container
4) container uploads photo to Nextcloud
5) photo moved to "Recently Deleted" on phone
6) delete syncs to iCloud.com
6) photo deleted by icloudpd from storage mounted in my container
7) photo remains in Nextcloud

This means that you can keep your phone/iCloud storage clear and access photos using the Nextcloud app

**Reply by CreativePR** [reply]
Understood and thank you for your explanation.  When I have some time, your container seems like a great option in lieu of my exploratory, test code I'm currently using.  Can you not get this added to the Nextcloud App Store?  I'm really not sure why such functionality is not already included as there is options for Google Drive/Photos, One Drive and Dropbox imports (but not syncs for some reason??)  Seems like a no brainer to include iCloud sync (or at least import functionality) as well.

So, the issue is that your container still leaves us at the same point.  With our users, they do not want to pay any paid level of iCloud storage and they do not want photos deleted from their camera roll until they choose to manually delete them.  At the very worst case scenerio, user might accept to have the oldest photos deleted to keep them within the 5GB limit (so we can continue to leverage iCloud for background loading to Nextcloud)

Thanks

**Comment by AndreyNikiforov** [comment]
> Problem to solve
> 
> * if command is sent to delete from iCloud by ICPD, then the photo is also deleted from the user's iCloud camera roll. As stated above our users do not want this behavior. They want the photos to remain in their camera roll until they choose to delete them even if they have been deleted from their iCloud account.

IIUC When you choose to synchronize iPhone camera roll with iCloud, then removing from either iPhone or iCloud, will delete in both places. That is Apple behavior and has nothing to do with icloudpd (icloudpd mimics/emulates web access to iCloud.com and cannot do anything that is not supported by icloud.com web interface). You can test that behavior yourself on any account.

**Reply by CreativePR** [reply]
Yes, this was tested and confirmed.  That why it is a problem to try and think of a way to solve.  How can it be possible somehow to work around this "Apple Behavior" so at a minimum, the maximum amount of photos will stay in the user's iCloud (and therefore not deleted from the user's device) to keep them just under the 5GB limit why they continue to generate new photos/videos going forward?

Thanks

**Reply by semiligneous** [reply]
Found this project while investigating this exact use case. Did anyone come up with any ideas/solutions to this?

If all photos get deleted locally from device (as a result of being deleted from iCloud) within a short period of time (minutes/hours presumably) then the problem to solve becomes 'how would the user be able to view (and send, share etc) their photos?'

The 'obvious' solution would be to have an alternate photo viewer, which works 'nicely' like inbuilt Photos. This would need to pull down from the user's backup solution (requiring the backup solution to either be publicly accessible or the user to be running a vpn/remote access solution), with a configuration able cache (measured in either size (gb) or time (data) ie 'keep 3 months of photos locally' or 'keep up to 20gb photos locally).

It's not really very elegant and I don't think I'd want to do it... but I'd be interested if there's anyone doing anything like that..?!

**Reply by boredazfcuk** [reply]
In my container, I have added the capability to upload files to a Nextcloud server. Syncing deletes is a setting which needs to be enabled.

So icloud-photos-downloader downloads all your pictures, uploads them to a Nextcloud server, then when you delete stuff from icloud, these deletions are not sync'd to Nextcloud.

This means you have to manually delete all the crap photos from Nextcloud, but it also means that's you can clear everything off your phone and the photos will remain on your Nextcloud server.

**Comment by semiligneous** [comment]
Right, so do you have almost no photos/videos stored in the Photos app on your phone?

If you want to show a photo to a friend or whatever, what do you do? Just open Nextcloud app and do it there instead?

How do you find that day to day? Photos is obviously a much smoother experience browsing photos than Nextcloud, and I worry that using only Nextcloud would be quite awkward?

----- Original message -----

**Comment by AndreyNikiforov** [comment]
One option to limit size of the icloud would be to sync using moving window, say, last 30 days, and delete from icloud anything older than 30days. That way local collection will grow, but icloud/photos will be limited to last 30days. "Moving window" can be extended to include favorites, or assets in albums, or any other criteria, to keep old, but valuable photos on devices and in icloud for easy access.

**Reply by semiligneous** [reply]
That could work, so you're saying it's possible to run a _move_ command (_--delete-after-download_) and pass a series of options so that it won't move files that are **either** older than X days **OR** not in album X, Y or Z ?

If so, that seems like it would work to keep the iCloud usage down to a sensible level (each user will be different to stay within the 5gb) while also leaving some recent and/or important photos on the device.

Am I understanding correctly?

**Reply by AndreyNikiforov** [reply]
I was validating if proposed solution solves your problem, but I did not clarify that proposed solution will need extra development, specifically, support for new filters.

**Reply by jhofker** [reply]
I would love this. Either keep the last x days or a feature that allows x GB of storage to remain in-use on icloud.

**Reply by jhofker** [reply]
Ah yeah, #659 is what I'd be looking for here. Just to link the request so others see it.

**Comment by semiligneous** [comment]
Got it, understood! Yes that would probably be a solution - if the features already existed I would be tempted to have a play with it and see if it worked. I’m not sure I would be comfortable asking you or someone to build it out though - I’m not 100% sure it’s what I want to do right now so don’t want to waste your time and resources on something I may not end up using!

----- Original message -----

**Comment by CreativePR** [comment]
Being able to leverage the free 5GB of iCloud that is given to every iPhone user would be very helpful in converting iPhone users to Nextcloud.  Right now Nextcloud iOS app for auto uploading works terribly.  Transferring photos to Nextcloud from iCloud with minimal user interactions is an ideal solution.

So yes, if something can be setup that will help the script keep as many photos as possible in iCloud (so they can be seen/worked with in the native iOS photos app) to keep it just under the 5GB limit, this would be great.  If a user needs something much older than what is in that current 5GB space (therefore in the iOS photos app) they could just open up their Nextcloud app and find the older content there.

The majority of iPhone users are typically dealing with more recent photos and not so often are looking for older content.

**Comment by AndreyNikiforov** [comment]
see [RFC](https://github.com/icloud-photos-downloader/icloud_photos_downloader/issues/1046) and related [`--keep-icloud-recent-days`](https://icloud-photos-downloader.github.io/icloud_photos_downloader/reference.html#keep-icloud-recent-days-parameter) parameter
